ompi/mpi/c/attr_fn_abi.c

```diff
@@ -0,0 +1,130 @@
+
+/*
+ * variables used for certain predefined attributes
+ * for windows
+ */
+int ompi_abi_mpi_win_flavor_create = MPI_WIN_FLAVOR_CREATE_ABI_INTERNAL;
+int ompi_abi_mpi_win_flavor_allocate = MPI_WIN_FLAVOR_ALLOCATE_ABI_INTERNAL;
+int ompi_abi_mpi_win_flavor_shared = MPI_WIN_FLAVOR_SHARED_ABI_INTERNAL;
+int ompi_abi_mpi_win_flavor_dynamic = MPI_WIN_FLAVOR_DYNAMIC_ABI_INTERNAL;
+int ompi_abi_mpi_win_model_unified = MPI_WIN_UNIFIED_ABI_INTERNAL;
+int ompi_abi_mpi_win_model_separate = MPI_WIN_SEPARATE_ABI_INTERNAL;
+
+/*
+ * predefined callbacks for win, comm, type attributes
+ */
+int ABI_C_MPI_TYPE_NULL_DELETE_FN( MPI_Datatype_ABI_INTERNAL datatype, int type_keyval,
+                                    void* attribute_val_out,
+                                    void* extra_state )
+{
+   /* Why not all MPI functions are like this ?? */
+   return MPI_SUCCESS;
+}
+
+int ABI_C_MPI_TYPE_NULL_COPY_FN( MPI_Datatype_ABI_INTERNAL datatype, int type_keyval,
+                                  void* extra_state,
+                                  void* attribute_val_in,
+                                  void* attribute_val_out,
+                                  int* flag )
+{
+   *flag = 0;
+   return MPI_SUCCESS;
+}
+
+int ABI_C_MPI_TYPE_DUP_FN( MPI_Datatype_ABI_INTERNAL datatype, int type_keyval,
+                            void* extra_state,
+                            void* attribute_val_in, void* attribute_val_out,
+                            int* flag )
+{
+   *flag = 1;
+   *(void**)attribute_val_out = attribute_val_in;
+   return MPI_SUCCESS;
+}
+
+int ABI_C_MPI_WIN_NULL_DELETE_FN( MPI_Win_ABI_INTERNAL window, int win_keyval,
+                                   void* attribute_val_out,
+                                   void* extra_state )
+{
+   return MPI_SUCCESS;
+}
+
+int ABI_C_MPI_WIN_NULL_COPY_FN( MPI_Win_ABI_INTERNAL window, int win_keyval,
+                                 void* extra_state,
+                                 void* attribute_val_in,
+                                 void* attribute_val_out, int* flag )
+{
+   *flag= 0;
+   return MPI_SUCCESS;
+}
+
+int ABI_C_MPI_WIN_DUP_FN( MPI_Win_ABI_INTERNAL window, int win_keyval, void* extra_state,
+                           void* attribute_val_in, void* attribute_val_out,
+                           int* flag )
+{
+   *flag = 1;
+   *(void**)attribute_val_out = attribute_val_in;
+   return MPI_SUCCESS;
+}
+
+int ABI_C_MPI_COMM_NULL_DELETE_FN( MPI_Comm_ABI_INTERNAL comm, int comm_keyval,
+                                    void* attribute_val_out,
+                                    void* extra_state )
+{
+   fprintf(stderr,"inside ABI_C_MPI_COMM_NULL_DELETE_FN\n");
+   return MPI_SUCCESS;
+}
+
+int ABI_C_MPI_COMM_NULL_COPY_FN( MPI_Comm_ABI_INTERNAL comm, int comm_keyval,
+                                  void* extra_state,
+                                  void* attribute_val_in,
+                                  void* attribute_val_out, int* flag )
+{
+   *flag= 0;
```

Comment:
Minor whitespace nit

```suggestion
   *flag = 0;
```

Reply:
fixed